Libraries/LibMedia/Audio/PlaybackStreamWasapi.cpp

```diff
@@ -0,0 +1,502 @@
+        DWORD result = WaitForMultipleObjects(2, handles.data(), FALSE, INFINITE);
+        switch (result) {
+        case WAIT_OBJECT_0: {
+            state->task_queue_mutex.lock();
+            auto task = state->task_queue.dequeue();
+            state->task_queue_mutex.unlock();
+            using enum Task::Kind;
+            switch (task.kind) {
+            case Play:
+                state->audio_client->Start();
+                task.resolve(total_time_played_impl(state));
+                state->playing = true;
+                break;
+            case DrainAndSuspend:
+                // FIXME: Make this do the drain bufffer part
+                state->audio_client->Stop();
+                state->playing = false;
+                task.resolve({});
+                break;
+            case DiscardAndSuspend:
+                state->audio_client->Stop();
+                state->audio_client->Reset();
+                state->playing = false;
+                task.resolve({});
+                break;
+            default:
+                VERIFY_NOT_REACHED();
+            }
+            DWORD res = WaitForSingleObject(handles[1], 0);
+            // Both the task event and buffer event were signaled
+            if (res == WAIT_OBJECT_0)
+                break;
+            // The buffer event wasn't signaled, so we skip the iteration
+            if (res == WAIT_TIMEOUT)
+                continue;
+            // Unless the wait errors we should never hit this
+            VERIFY_NOT_REACHED();
+        }
+        case WAIT_OBJECT_0 + 1:
+            break;
+        default:
+            VERIFY_NOT_REACHED();
+        }
+
+        // We check that after the wait we weren't asked to exit.
+        if (state->exit_requested.load(MemoryOrder::memory_order_acquire))
+            break;
+
+        if (!state->playing)
+            continue;
+
+        u32 padding = 0;
+        hr = state->audio_client->GetCurrentPadding(&padding);
+        // TODO: Have sane action on failing. This is for testing only.
+        if (FAILED(hr)) {
+            dbgln("GetCurrentPadding failed with: {}", Error::from_windows_error(hr));
+            VERIFY_NOT_REACHED();
+        }
+        u32 frames_available = state->buffer_frame_count - padding;
+
+        if (frames_available == 0)
+            continue;
+        BYTE* buffer;
+        hr = state->render_client->GetBuffer(frames_available, &buffer);
+        if (FAILED(hr)) {
+            dbgln("GetBuffer failed with: {}", Error::from_windows_error(hr));
+            VERIFY_NOT_REACHED();
+        }
+
+        u32 buffer_size = frames_available * block_align;
+        ReadonlyBytes bytes_written = state->data_request_callback({ buffer, buffer_size }, PcmSampleFormat::Float32, frames_available);
+        if (bytes_written.is_empty()) {
+            if (state->underrun_callback)
+                state->underrun_callback();
+            buffer_flags |= AUDCLNT_BUFFERFLAGS_SILENT;
```

Comment:
With `buffer_flags` at the function scope, won't this cause the audio to remain silent permanently after the first underrun?